galley/pkg/config/processor/transforms/serviceentry/converter/instance.go

```diff
@@ -0,0 +1,253 @@
+		return
+	}
+
+	spec := service.Item.(*coreV1.ServiceSpec)
+
+	resolution := networking.ServiceEntry_STATIC
+	location := networking.ServiceEntry_MESH_INTERNAL
+	endpoints := convertExternalServiceEndpoints(spec, service.Metadata)
+
+	// Check for an external service
+	externalName := ""
+	if spec.Type == coreV1.ServiceTypeExternalName && spec.ExternalName != "" {
+		externalName = spec.ExternalName
+		resolution = networking.ServiceEntry_DNS
+		location = networking.ServiceEntry_MESH_EXTERNAL
+	}
+
+	// Check for unspecified Cluster IP
+	addr := model.UnspecifiedIP
+	if spec.ClusterIP != "" && spec.ClusterIP != coreV1.ClusterIPNone {
+		addr = spec.ClusterIP
+	}
+	if addr == model.UnspecifiedIP && externalName == "" {
+		// Headless services should not be load balanced
+		resolution = networking.ServiceEntry_NONE
+	}
+
+	ports := make([]*networking.Port, 0, len(spec.Ports))
+	for _, port := range spec.Ports {
+		ports = append(ports, convertPort(port))
+	}
+
+	host := serviceHostname(service.Metadata.Name, i.domain)
+
+	// Store everything in the ServiceEntry.
+	out.Hosts = []string{host}
+	out.Addresses = []string{addr}
+	out.Resolution = resolution
+	out.Location = location
+	out.Ports = ports
+	out.Endpoints = endpoints
+	out.ExportTo = convertExportTo(service.Metadata.Annotations)
+
+	// Convert Metadata
+	outMeta.Name = service.Metadata.Name
+	outMeta.Labels = service.Metadata.Labels.Clone()
+
+	// Convert the creation time.
+	outMeta.CreateTime = service.Metadata.CreateTime
+
+	// Update the annotations.
+	outMeta.Annotations = service.Metadata.Annotations.CloneOrCreate()
+
+	// Add an annotation for the version of the service resource.
+	outMeta.Annotations[annotations.ServiceVersion] = string(service.Metadata.Version)
+}
+
+func convertExportTo(annotations resource.StringMap) []string {
+	var exportTo map[string]struct{}
+	if annotations[kube.ServiceExportAnnotation] != "" {
+		exportTo = make(map[string]struct{})
+		for _, e := range strings.Split(annotations[kube.ServiceExportAnnotation], ",") {
+			exportTo[strings.TrimSpace(e)] = struct{}{}
+		}
+	}
+	if exportTo == nil {
+		return nil
+	}
+
+	out := make([]string, 0, len(exportTo))
+	for k := range exportTo {
+		out = append(out, k)
+	}
+	sort.Strings(out)
+	return out
+}
+
+// convertEndpoints applies the k8s Endpoints to the output.
+func (i *Instance) convertEndpoints(endpoints *resource.Entry, outMeta *resource.Metadata, out *networking.ServiceEntry) {
+	if endpoints == nil {
+		return
+	}
+
+	spec := endpoints.Item.(*coreV1.Endpoints)
+	// Store the subject alternate names in a set to avoid duplicates.
+	subjectAltNameSet := make(map[string]struct{})
+	eps := make([]*networking.ServiceEntry_Endpoint, 0)
+
+	// A builder for the annotation for not-ready addresses. The ServiceEntry does not support not ready addresses,
+	// so we send them as an annotation instead.
+	var notReadyBuilder strings.Builder
```

Comment:
Maybe we should add support for string builders in pkg/pool?

Reply:
Added a TODO to do this after merge to master.